### PR TITLE
Desktop: Fixes #5686: Fixed Tags Order

### LINKED ISSUE #5686: Tags order
The tag order between left panel and windows tag is different.

tag order, PNG

## Environment
Joplin version: 2.5.10
Platform: Windows 
OS specifics: Win10

## Steps to reproduce
1. Create tags with emoji, #, @, ...
2. Open tags windows, and check tags order in the drop-down menu

## Describe what you expected to happen
I expect the same order in the drop-down menu than the tag panel.

### PR DESCRIPTION
**Pull request by OmGole**
This fixes the bug #5686; The tag order between left panel and windows tag is different.
tags order
Whenever tags starting with emoji, #, @ were added then the order for tags in the promptDialog dropdown menu and sidebar TAGS section were different.

The cause of the bug was that different sorting logic was used in the sidebar tags section and setTags.ts (the sort method in this file is used to sort tags in the promptDialog dropdown menu).

To fix this I did the following, removed the old sort method in setTags.ts and added the sort method which was used in the side-menu-shared.js (this sort method is used to sort tags in the sidebar). As the sort methods are same the outputs are also the same. I have also added the unit test.

### CODE REVIEW

File: packages/lib/models/Tag.test.js [hunk 1]
```diff
@@ -132,28 +173,109 @@ describe('models/Tag', function() {
...
+		commonTagIds = commonTags.map((t) => t.id);
 		expect(commonTags.length).toBe(3);
 		expect(commonTagIds.includes(taga.id)).toBe(true);
 		expect(commonTagIds.includes(tagb.id)).toBe(true);
 		expect(commonTagIds.includes(tagc.id)).toBe(true);
-	}));
+	});
 
+	it('should sort tags in promptDialog', async () => {
+		const folder1 = await Folder.save({ title: 'folder1' });
```
Comment (roman-r-m) [inline packages/lib/models/Tag.test.js:206]: This test does not cover your changes at all.
Sure you copied the logic from `setTags` here but now if someone will accidentally breaks `setTags`, your test will still be passing. Which defeats the whole purpose of it.

Comment (OmGole) [reply packages/lib/models/Tag.test.js:206]: Then how do I test it? I will have to first export it using global variables, right? I am actually writing my first test here so just asking for feedback?

Comment (roman-r-m) [reply packages/lib/models/Tag.test.js:206]: Your test needs to call `setTags`

Comment (OmGole) [reply packages/lib/models/Tag.test.js:206]: I will write a new test accordingly, I just need a little push, is exporting a variable from a file for testing allowed? I don't see another way to test without exporting?

Comment (OmGole) [reply packages/lib/models/Tag.test.js:206]: But won't that test the entire file. I mean I just want to test the startTags and tagSuggestions from the entire file.

Comment (roman-r-m) [reply packages/lib/models/Tag.test.js:206]: If you don't want to test the whole file (which would be useful btw) you can extract the parts you want to test to separate methods and test them.

Comment (OmGole) [reply packages/lib/models/Tag.test.js:206]: Okay I will try to test the entire file. Also, extracting the parts using methods will work on classes but setTags uses functions. I researched about it on stackoverflow and found out that in order to export a variable from a function we will have to use global variables.

**Review by CalebJohn — CHANGES_REQUESTED** [review]

**Comment by OmGole** [comment]
Update: I am planning to wrap the sorting logic in a function so that it can be used to sort tags anywhere in the codebase. This will help us maintain consistency in the sorting logic everywhere and also maintain cleaner code. Wrapping it in a function will also help me export it to a test file and perform unit tests on it. Also thanks to @roman-r-m for all the constructive feedback!!! 

Here sorting logic while be implemented using localeCompare as stated above by @CalebJohn

**Comment by laurent22** [comment]
It looks like the formatting of the code has been changed by prettier or something. Please revert these cosmetic changes and keep only your own changes.

**Comment by OmGole** [comment]
Ok

**Comment by OmGole** [comment]
Hey @laurent22 my changes  will consist of more than 50 lines of code should I go ahead and do it or make a post about it on the joplin forum?

**Comment by OmGole** [comment]
Update: I have reverted back the changes made by prettier. I have added a static method (sortTags) in Tag.ts to sort tags. This method will help us to maintain consistency in sorting tags anywhere in the codebase on top of that it will also allow us to make changes to the sorting functionality less cumbersome. If we have to change the sorting functionality all we have to do is change the sortTags method. I used the sortTags method in three files i.e setTags.ts (used for promptDialog default values and promptDialog dropdown values), Taglist.tsx (used for showing tags in the bottom left corner of the noteEditor), side-menu-shared.js (used for showing tags on the sidebar). Also I have written test for the sortTags method in Tag.test.js

File: packages/lib/models/Tag.ts [hunk 2]
```diff
@@ -216,4 +216,27 @@ export default class Tag extends BaseItem {
 			return tag;
 		});
 	}
+
+	static sortTags(tags: any) {
```
Comment (roman-r-m) [inline packages/lib/models/Tag.ts:220]: Looks like you're using this function to compare objects of different types in different places: sometimes it's tags, other times it's `{ value: tag.id, label: tag.title }`
It seems to me it's best to rewrite it to always have args of the same time. Then you can also get rid of `any` and give it proper type.

Comment (OmGole) [reply packages/lib/models/Tag.ts:220]: sortTags sorts on the basis of the title of the tag so it will work for `{ value: tag.id, label: tag.title }` and tags as well. I can write an if statement for it in sortTags but even without it the functionality will work.

Comment (roman-r-m) [reply packages/lib/models/Tag.ts:220]: The issue with how it's written right now is that if someone wanted to use your function elsewhere, it's not possible to understand by looking at the signature what kind parameters it expects. You have to look at its implementation to find that it should be an array of things that have either title or label.

At the same time, you can make a trivial change to specify that this function requires an array of Tags and it makes things much more clear.

Comment (laurent22) [reply packages/lib/models/Tag.ts:220]: For the type, it's possible to use TagEntity. Also I think this function shouldn't support sorting by label. Instead the view should be changed so instead of having Load tags => convert to list items => sort. You should have Load tags => sort => convert to list items. That way you don't need any special case in this function

**Comment by OmGole** [comment]
Update: I have added type to tags in `sortTags` as requested and also removed the if statement for label as suggested by laurent

File: packages/lib/models/Tag.ts [hunk 3]
```diff
@@ -216,4 +216,20 @@ export default class Tag extends BaseItem {
 			return tag;
 		});
 	}
+
+	static sortTags(tags: TagEntity[]) {
+		if (tags[0].title) {
```
Comment (roman-r-m) [inline packages/lib/models/Tag.ts:221]: I don't think you need the condition anymore

Comment (OmGole) [reply packages/lib/models/Tag.ts:221]: oops!! will remove it asap

File: packages/lib/models/Tag.ts [hunk 4]
```diff
@@ -216,4 +216,18 @@ export default class Tag extends BaseItem {
 			return tag;
 		});
 	}
+
+	static sortTags(tags: TagEntity[]) {
+		return tags.sort((a: any,b: any)=>{
```
Comment (laurent22) [inline packages/lib/models/Tag.ts:221]: still need types here

File: packages/lib/models/Tag.test.js [hunk 5]
```diff
@@ -156,4 +156,38 @@ describe('models/Tag', function() {
...
+		// defaultTags contains only the id and title of all the unsorted tags
+		const defaultTags = commonTags.map(a => {
+			return { value: a.id, label: a.title };
+		});
+
+		// sortedTagsTitles contains the titles of all the sorted tags
+		const sortedTagsTitles = Tag.sortTags(defaultTags).map(a => a.label);
+
+		expect(sortedTagsTitles).toEqual(['@⏲15 min', '@⏲30 min', '@⏲60 min', '#coding', '#house', '#wait']);
+		expect(sortedTagsTitles).not.toEqual(['@⏲15 min', '#house','#coding', '@⏲60 min', '#wait', '@⏲30 min']);
```
[the remark this answers is not in the document]
Comment (laurent22) [reply packages/lib/models/Tag.test.js:191]: But the logic... If a = 1, do you think you need to test that a != 2?

File: packages/lib/models/Tag.test.js [hunk 6]
```diff
@@ -156,4 +156,38 @@ describe('models/Tag', function() {
...
+		await Tag.addNote(tag4.id, note1.id);
+		await Tag.addNote(tag2.id, note1.id);
+		await Tag.addNote(tag1.id, note1.id);
+		await Tag.addNote(tag6.id, note1.id);
+		await Tag.addNote(tag3.id, note1.id);
+		await Tag.addNote(tag5.id, note1.id);
+
+		// This test checks the sorting functionality for tags in all three files(setTags, Sidebar and TagList)
+		// commonTags contains unsortedTags
+		const commonTags = await Tag.commonTagsByNoteIds([note1.id]);
```
Comment (laurent22) [inline packages/lib/models/Tag.test.js:180]: That doesn't do what you say it's doing. With the way it's implemented now, you don't need to test setTags or the SideBar, so please remove this test.

File: packages/lib/models/Tag.test.js [hunk 7]
```diff
@@ -156,4 +156,38 @@ describe('models/Tag', function() {
...
+
+		// defaultTags contains only the id and title of all the unsorted tags
+		const defaultTags = commonTags.map(a => {
+			return { value: a.id, label: a.title };
+		});
+
+		// sortedTagsTitles contains the titles of all the sorted tags
+		const sortedTagsTitles = Tag.sortTags(defaultTags).map(a => a.label);
```
Comment (laurent22) [inline packages/lib/models/Tag.test.js:188]: There shouldn't be anything about labels here.

File: packages/lib/models/Tag.test.js [hunk 8]
```diff
@@ -156,4 +156,38 @@ describe('models/Tag', function() {
...
+		const tag4 = await Tag.save({ title: '@⏲15 min' });
+		const tag5 = await Tag.save({ title: '@⏲30 min' });
+		const tag6 = await Tag.save({ title: '@⏲60 min' });
+
+		await Tag.addNote(tag4.id, note1.id);
+		await Tag.addNote(tag2.id, note1.id);
+		await Tag.addNote(tag1.id, note1.id);
+		await Tag.addNote(tag6.id, note1.id);
+		await Tag.addNote(tag3.id, note1.id);
+		await Tag.addNote(tag5.id, note1.id);
```
Comment (laurent22) [inline packages/lib/models/Tag.test.js:176]: You don't need any of this. Just mock the data - `[{ title:'#coding' }, { title:'#wait' }, etc.`

File: packages/lib/models/Tag.test.js [hunk 9]
```diff
@@ -156,4 +156,38 @@ describe('models/Tag', function() {
...
+		const defaultTags = commonTags.map(a => {
+			return { value: a.id, label: a.title };
+		});
+
+		// sortedTagsTitles contains the titles of all the sorted tags
+		const sortedTagsTitles = Tag.sortTags(defaultTags).map(a => a.label);
+
+		expect(sortedTagsTitles).toEqual(['@⏲15 min', '@⏲30 min', '@⏲60 min', '#coding', '#house', '#wait']);
+		expect(sortedTagsTitles).not.toEqual(['@⏲15 min', '#house','#coding', '@⏲60 min', '#wait', '@⏲30 min']);
+	}));
```
Comment (laurent22) [inline packages/lib/models/Tag.test.js:192]: Please add more tests - empty list, tags with no title property, etc. Read the code you've copied and pasted in sortTags and make sure you test all cases.

**Comment by laurent22** [comment]
Please don't mark comments as "resolved" because then we don't know what changes have been reviewed or not.

**Comment by OmGole** [comment]
Hey @laurent22, I have unresolved the previous conversations can you please review it again?

**Comment by laurent22** [comment]
All good now, thanks @OmGole for adding the tests and making these additional changes.
